### DISCUSSION TITLE
Scene related. Enum component can not be reflected.

### DISCUSSION THREAD

**Discussion by automeowtion** [bevyengine/bevy#4038] [Q&A] [answered]
Is it correct to believe that enums can not be component and reflected? And wrapping enums with struct is the solution?

DETAIL:

In Reflection cleanup [#1536](https://github.com/bevyengine/bevy/pull/1536), @Davier mentioned that

> #[reflect_value(...)] should not be used for structs, they need #[reflect(...)] instead. I have no idea what happens when using both attributes on the same type, but I don't think it's supposed to happen.

From what I gathered: 
1. structs need `#[reflect(...)]`
2. enums need `#[reflect_value(...)]`
3. `#[reflect(...)]` and `#[reflect_value(...)]` don't mix
4. for a type to be used as component and reflected it needs `#[derive(Component)]` and `#[reflect(Component)]`

Which seems to mean that enums can not be used as components and reflected at the same time, because according to point number 3, `#[reflect(Component)]` and `#[reflect_value(...)]` can not co-exist. To test the above, I save a scene with an enum with the following derives
```
#[derive(Component, Serialize, Deserialize, PartialEq, Eq, Hash, Clone, Copy, Reflect)]
#[reflect_value(Serialize, Deserialize, PartialEq, Hash)]
#[reflect(Component)]
```
which causes the following panic:
> [...type] does not support ReflectValue serialization

What works: remove `#[reflect(Component)]`, and wrap the enum in a struct `pub struct EnumTWrapper(pub EnumT);`, which is then inserted as component.

**Comment by hymm** [comment]
There's a draft PR to add reflection to enums here: https://github.com/bevyengine/bevy/pull/1347.  Not sure about the status.

**Comment by Davier** [comment, marked as answer]
You can use `#[reflect_value(Component)]` on enums:
```rust
#[derive(Component, Serialize, Deserialize, PartialEq, Eq, Hash, Clone, Copy, Reflect)]
#[reflect_value(Component, Serialize, Deserialize, PartialEq, Hash)]
```

**Reply by alice-i-cecile** [reply]
Do we still need #1347 then?

**Reply by Davier** [reply]
Yes. That PR will allow to reflect what is inside an enum using `#[reflect()]`, like we already do for a reflected struct's fields.
Currently we have to use `#[reflect_value()]`, which treats the enums like black boxes (that´s why they need to implement `Serialize` and `Deserialize`, while structs don't).
